Create unit tests for validation

zachclark-ccpo opened the pull request:
## Overview
Using `jest` to test the business logic of the API endpoints.
Moved all of the validation to `utils/validation.ts`, each microservice will use these validation fns.
`npm test` output:
```
zclark@EDC-H6K8BH2:~/test-kyle-pr/atat-web-api/applications/portfolioDrafts$ npm test

> portfoliodrafts@0.1.0 test
> jest --config=jest.config.js

 PASS  test/requestbody.test.ts
  Testing validation of request body
    ✓ should return true because request body is present (2 ms)
    ✓ should return false because request body is empty JSON (1 ms)
    ✓ should return false when body is empty object (1 ms)
    ✓ should return false because request body is an empty String (1 ms)
    ✓ should return false because request body is empty whitespace String
    ✓ should return false because request body is null
    ✓ should return false because request body JSON is invalid (2 ms)
    ✓ should return true because request body JSON is valid
  Validation tests for createPortfolioStep function
    ✓ should map body to portfolioStep object
    ✓ should fail to map body to portfolioStep object due to missing attribute (2 ms)
    ✓ should fail to map body to portfolioStep object because request body is null
  Testing validation of path parameter
    ✓ should return false because path parameter is an empty string (1 ms)
    ✓ should return true because path parameter is a valid string (1 ms)
    ✓ should return false because path parameter is undefined (1 ms)

---------------|---------|----------|---------|---------|-------------------
File           | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
---------------|---------|----------|---------|---------|-------------------
All files      |     100 |      100 |     100 |     100 |                   
 validation.ts |     100 |      100 |     100 |     100 |                   
---------------|---------|----------|---------|---------|-------------------
Test Suites: 1 passed, 1 total
Tests:       14 passed, 14 total
Snapshots:   0 total
Time:        2.934 s, estimated 3 s
Ran all test suites.
```
### TODO

- [x] Ensure all files are being collected with code coverage (Im certain that some of the fns outside of validation.ts).
- [x] Split up tests into smaller files
- [x] Fix lerna errors

zachclark-ccpo committed:
Add commands.ts file, used to store dynamodb command files

First one added is portfolioStepCommand, placing it into its own file
so we can unit test the creation of the commands and test them against
a mock dynamo table

zachclark-ccpo committed:
Baseline unit tests for validation and createPortfolioStep

unit tests for various validation methods, aswell as object
mapping for the portfolioStep object

zachclark-ccpo committed:
Add pathParamter validation and request body validation

- isBodyPresent boolean valdiation to check for valid JSON but empty
request body
- isPathParameterPresent, same thing but for the path parameter

zachclark-ccpo committed:
Update createPortfolioStep to use new validation

now using isPathParameterPresent and isBodyPresent

zachclark-ccpo committed:
Update package.json

- add jest test script
- add jest dependency
- add ts-jest dependency

zachclark-ccpo committed:
Add jest config file to applications/portfolioDrafts

Used for jest tests, matches any file with '.test.ts' file extension

zachclark-ccpo committed:
Update isBodyPresent to improve code coverage

zachclark-ccpo committed:
Remove complexity from tests

zachclark-ccpo committed:
Enable code coverage in jest.config

zachclark-ccpo commented:
Lerna errors:
```
npm ERR! Missing: @types/jest@^26.0.24
npm ERR! Missing: jest@^27.0.6
npm ERR! Missing: ts-jest@^27.0.4
npm ERR! 
```

kylelaker-ccpo commented on applications/portfolioDrafts/utils/commands.ts:
```diff
@@ -0,0 +1,27 @@
+import { PortfolioStep } from "../models/PortfolioStep";
+
+export function portfolioStepCommand(
+  table: string | undefined,
```
This doesn't handle undefined, it'd be invalid to pass undefined to the API, and so we should not allow undefined as a type.

kylelaker-ccpo commented on applications/portfolioDrafts/utils/validation.ts:
```diff
@@ -15,6 +16,30 @@ export function isValidJson(str: string): boolean {
+ * @param body - The body of the request
+ * @returns true if the string is empty or null
+ */
+export function isBodyPresent(body: string | null): body is string {
```
This is may be the wrong use of a type guard. It does accomplish the goals we have and it does in fact assert that the input is a string. But it fails to capture the additional restrictions imposed within the body.

zachclark-ccpo committed:
Change portfolioStepCommand to only accept string

Before: string | undefined
Changed to: string
The actual UpdateCommand will not handle an undefined table name,
so it must be removed from the portfolioStepCommand

zachclark-ccpo committed:
Update package-lock.json file using correct npm version

zachclark-ccpo committed:
Merge branch 'develop' into AT-6415

zachclark-ccpo commented on applications/portfolioDrafts/jest.config.js:
```diff
@@ -0,0 +1,10 @@
+    transform: {
+      '^.+\\.tsx?$': 'ts-jest'
+    },
+    collectCoverage: true,
```
There is a function invoked createPortfolioStep.ts, `const command = portfolioStepCommand(TABLE_NAME, portfolioDraftId, portfolioStep);` and some others that are not being included in the code coverage report. 
```
collectCoverageFrom: [
"**/*.ts"
]
```
yields the same results --> only validtions.ts is tested.
`collectCoverageFrom=<glob>` uses `glob` which I don't have the most experience with. Docs here: https://jestjs.io/docs/cli#--collectcoveragefromglob

zachclark-ccpo replied:
This is now resolved.

zachclark-ccpo committed:
Update gitignore to ignore coverage folder

zachclark-ccpo committed:
Merge branch 'AT-6415' of github.com:dod-ccpo/atat-web-api into AT-6415

zachclark-ccpo committed:
Merge branch 'AT-6415' of github.com:dod-ccpo/atat-web-api into AT-6415

zachclark-ccpo committed:
Create separate async command for createPortfolioStep

This will be used for unit testing locally with aws-sdk-mock

zachclark-ccpo committed:
Unit tests for UpdateCommand createPortfolioStepCommand

Not as useful as other mock tests, because we are not too concerned
if the object actually updates. But this proves a simple implementation of
local mocking.

zachclark-ccpo committed:
Move dynamodb call to its own function

This will allow us to unit test the getPortfolioStep dynamodb call,
and improves function clarity

kylelaker-ccpo commented on applications/portfolioDrafts/portfolio/createPortfolioStep.ts:
```diff
@@ -1,12 +1,14 @@
+import { isPortfolioStep, isValidJson, isBodyPresent, isPathParameterPresent } from "../utils/validation";
 
-const TABLE_NAME = process.env.ATAT_TABLE_NAME;
+const TABLE_NAME = process.env.ATAT_TABLE_NAME ?? "";
```
If this isn't set, we need to totally abort. We can't work with a default value. This would be an unexpected infrastructure issue. We probably want a helper that does something like:

```
if (!process.env.ATAT_TABLE_NAME) {
  throw new Error("Oh no! The table name isn't set.")
}
```

of course with a better message and maybe a subclass of `Error`? But this really is an exceptional case that should never happen and if it does, we need to return a 500 response with a really good log message ASAP.

kylelaker-ccpo commented on .gitignore:
```diff
@@ -3,6 +3,7 @@
 *.d.ts
 node_modules
 lerna-debug.log
+applications/portfolioDrafts/coverage
```
```suggestion
coverage/
```

zachclark-ccpo committed:
Update unit test for createPortfolioStep

- deleted unused imports
- refined test data

zachclark-ccpo committed:
Create unit test for getPortfolioStep

- mock GetCommand for getPortfolioStep
- this ensures that getPortfolioStepCommand is a proper
GetCommand (returns Item)

zachclark-ccpo committed:
Remove unused imports from requestbody.test.ts

zachclark-ccpo committed:
Update gitignore to follow team best practices
